File: leetcode/leetcode_py/rotting-oranges.py
'''
https://leetcode.com/problems/rotting-oranges

1. BFS
time: o(n)
space: o(n)
- Add all rotten orange in queue
- Visit adj nodes and mark it. After loop of queue size, add day += 1
- Check again Fresh orange exist
'''
class Solution:
    def orangesRotting(self, grid: List[List[int]]) -> int:
        q = []
        hasFresh = False
        for i in range(len(grid)):
            for j in range(len(grid[0])):
                if grid[i][j] == 2:
                    q.append((i,j))
                elif grid[i][j] == 1:
                    hasFresh = True
        if not hasFresh: return 0
        if not q: return -1

        res = 0
        while q:
            qSize = len(q)
            for k in range(qSize):
                i, j = q[0][0], q[0][1]
                q.pop(0)
                if i + 1 < len(grid) and grid[i+1][j] == 1:
                    grid[i+1][j] = 2
                    q.append((i+1, j))
                if j + 1 < len(grid[0]) and grid[i][j+1] == 1:
                    grid[i][j+1] = 2
                    q.append((i, j+1))
                if i > 0 and grid[i-1][j] == 1:
                    grid[i-1][j] = 2
                    q.append((i-1, j))
                if j > 0 and grid[i][j-1] == 1:
                    grid[i][j-1] = 2
                    q.append((i, j-1))
            if q:
                res += 1

        for i in range(len(grid)):
            for j in range(len(grid[0])):
                if grid[i][j] == 1: return -1

        return res

File: leetcode/leetcode_cpp/rotting-oranges.cpp
/*
https://leetcode.com/problems/rotting-oranges

1. BFS
time: o(n)
space: o(n)
- Add all rotten orange in queue
- Visit adj nodes and mark it. After loop of queue size, add day += 1
- Check again Fresh orange exist
*/
class Solution {
public:
    int orangesRotting(vector<vector<int>>& grid) {
        // BFS
        queue<pair<int, int>> q;
        
        bool noFresh = true;
        for (int i=0; i<grid.size(); ++i) {
            for (int j=0; j<grid[0].size(); ++j) {
                if (grid[i][j] == 2)
                    q.push({i, j});
                else if (grid[i][j] == 1)
                    noFresh = false;
            }
        }
        if (noFresh) return 0;
        if (q.empty()) return -1;

        int res = 0;
        while (!q.empty()) {
            int qSize = q.size();
            for (int k=0; k<qSize; ++k){
                int i = q.front().first, j = q.front().second; q.pop();
                if (i+1 < grid.size() && grid[i+1][j] == 1) {
                    grid[i+1][j] = 2;
                    q.push({i+1, j});
                }
                if (j+1 < grid[0].size() && grid[i][j+1] == 1) {
                    grid[i][j+1] = 2;
                    q.push({i, j+1});
                }
                if (i > 0 && grid[i-1][j] == 1) {
                    grid[i-1][j] = 2;
                    q.push({i-1, j});
                }
                if (j > 0 && grid[i][j-1] == 1) {
                    grid[i][j-1] = 2;
                    q.push({i, j-1});
                }
            }
            if (!q.empty()) res += 1;
        }
        for (int i=0; i<grid.size(); ++i) {
            for (int j=0; j<grid[0].size(); ++j) {
                if (grid[i][j] == 1)
                    return -1;
            }
        }
        return res;
    }
};